enhancements/20220823-node-manager-support-ksm.md

```diff
@@ -0,0 +1,265 @@
+
+```yaml
+spec:
+  run: run
+  mode: customized
+  thresCoef: 20
+  ksmtuneddParameters:
+    boost: 300
+    decay: 50
+    maxPages: 2000
+    minPages: 100
+    sleepMsec: 20
+```
+
+When `spec.run` is `stop`, it will suspend running ksmtuned, but continue to keep the shared pages.
+
+```yaml
+apiVersion: harvesterhci.io/v1beta1
+kind: Ksmtuned
+metadata:
+  name: harvester-master
+spec:
+  run: stop
+  mode: standard
+  thresCoef: 20
+  ksmtunedParameters:
+    boost: 0
+    decay: 0
+    maxPages: 100
+    minPages: 100
+    sleepMsec: 20
+status:
+  sharing: 17299179
+  shared: 312514
+  unshared: 2334736
+  volatile: 65571 
+  fullScans: 14102
+  stableNodeChains: 30
+  stableNodeDups: 56249
+  ksmdPhase: Stopped
+```
+
+When `spec.run` is `prune`, it will suspend ksmtuned and clean up the shared memory pages.
+
+```yaml
+apiVersion: harvesterhci.io/v1beta1
+kind: Ksmtuned
+metadata:
+  name: harvester-master
+spec:
+  run: prune
+  mode: standard
+  thresCoef: 20
+  ksmtunedParameters:
+    boost: 0
+    decay: 0
+    maxPages: 100
+    minPages: 100
+    sleepMsec: 20
+status:
+  sharing: 0
+  shared: 0
+  unshared: 0
+  volatile: 0 
+  fullScans: 0
+  stableNodeChains: 0
+  stableNodeDups: 0
+  ksmdPhase: Pruned
+```
+
+Ksmtuned CR’s status to display [KSM metrics](https://www.kernel.org/doc/html/latest/admin-guide/mm/ksm.html#ksm-daemon-sysfs-interface) and ksmd phase.
+
+```yaml
+apiVersion: harvesterhci.io/v1beta1
+kind: Ksmtuned
+metadata:
+  name: harvester-master
+spec:
+  ...
+status:
+  sharing: 17299179
+  shared: 312514
+  unshared: 2334736
+  volatile: 65571 
+  fullScans: 14102
+  stableNodeChains: 30
+  stableNodeDups: 56249
+  ksmdPhase: Stopped/Running/Pruned
+```
+
+**Ksmtuned script**
+
+Since KSM runs in a single mode, it is possible for large memory devices to occupy the CPU for a long time. To solve this problem, you can set the ksmtuned script threshold to run KSM only when the free memory is less than a certain percentage.
+
+`Ksmtuned.spec` parameters：
+
+- **run:** `stop`/`run`/`prune`, stop/run Ksmtuned or prune share pages。
+- **mode:** `standard`/`high`/`customized`.
+- **thresCoef**: Free memory threshold, KSM starts when the memory available is less than the set percentage ratio, otherwise KSM stops.
+- **ksmtunedParameters.boost**: Increment the number of pages scanned at a time, incrementing if the memory availability is less than the threshold.
+- **ksmtunedParameters.decay**: Decrement the number of pages scanned at a time, decrementing if memory availability is greater than the threshold.
+- **ksmtunedParameters.maxPages**: Maximum number of pages per scan.
+- **ksmtunedParameters.minPages**: The minimum number of pages per scan, i.e. the configuration for the first run.
+- **ksmtunedParameters.sleepMsec**: Each time you change the configuration interval, calculate the interval between two scans by the formula.(`sleepMsec * 16 * 1024 * 1024 / total memory` ).
```

Comment:
LGTM. Thanks for the well-written HEP.
nit: It will be nice to have an example (or maybe in the document) of how these parameters work.
For example, what does the controller do if the user gives:
```
spec:
  ksmtunedParameters:
    boost: 200
    decay: 50
    maxPages: 10000
    minPages: 100
    sleepMsec: 20
  mode: high
  run: run
  thresCoef: 80
```

I need to read the code to understand the controller increase the `pages_to_scan ` value up to `maxPages` value every minute if the free memory is less than the threshold and gradually decreases `pages_to_scan` value to `minPages` if memory is freed.

Reply:
It will be written in a document.